Add normalize_and_get_length() to Vector2 and Vector3

lawnjelly opened the pull request:
Length is calculated as part of the normalize routine. This is often desired, and also can be used to flag the condition where the Vector was too small to normalize (by returning zero length).

Helps address #74852

## Notes
Most engines I've worked on have returned the length from normalize, as this can be a useful value in a lot of cases, and should cost nothing from c++ if not used (as it should be optimized out).

* In a PR meeting it was decided this would be better as a separate function, rather than modifying the existing `normalize()`.
* Returning zero if less then EPSILON allows users to deal with error cases where the input Vector length is small.
* Due to the ability to handle error conditions (over the existing `normalized()` bound function) the function is bound, in order to be available from script etc.
* I noticed that `Vector3::normalize()` sets all values to zero if length is zero (curiously `Vector2` does not do this). I'm presuming this is to get a standard value for later comparisons because of the different bit representations near zero. This didn't seem appropriate to do in this PR with an epsilon, but can change if desired.

This can be done longhand which is what I normally end up doing, e.g.:
```
real_t length = myvec.length();
if (length >= EPSILON)
{
myvec /= length;
}
// do something with length
```
but the form enabled by the PR is more concise and less error prone (especially epsilons can be standardized in the normalize function):
```
real_t length = myvec.normalize_and_get_length();
if (length == 0)
{
// Failed ...
}
// do something with length
```
or simply
```
if (myvec.normalize_and_get_length() == 0)
{
// Failed
}
```
_(This is one of those rare occasions we can do an exact floating point comparison)_
<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->

aaronfranke commented on core/math/vector2.h:
```diff
@@ -251,6 +251,17 @@ _FORCE_INLINE_ bool Vector2::operator!=(const Vector2 &p_vec2) const {
 }
 
+real_t Vector2::normalize() {
+	real_t l = x * x + y * y;
+	if (l != 0) {
+		l = Math::sqrt(l);
+		x /= l;
+		y /= l;
+		return l;
+	}
+	return 0.0;
```
Could be simplified:

```suggestion
	real_t l = x * x + y * y;
	if (l != 0) {
		l = Math::sqrt(l);
		x /= l;
		y /= l;
	}
	return l;
```

lawnjelly replied:
Yup sure. :+1: 
I was intrigued to find in godbolt that the assembly is slightly different.

Even adjusting for the literal issue I mentioned in #56396 (and is probably addressed in a separate PR if we get some agreement on it), I wonder if it is trying to account for signed zero. We'd have to time them to see which was "better", I don't expect there's a lot in it! :grin:

aaronfranke replied:
`-0.0` would be the only case that is different, so that has to be it. I also checked NaN but that should enter the `if` and then explode things there so it wouldn't be effected by this.

aaronfranke reviewed (commented):
Note: This would only be for internal use. `normalized` is what's exposed, and it returns a `Vector2`.

lawnjelly commented:
PR Meeting: Add `normalize_and_get_length()` separate function if this has enough use, rather than change `normalize()`.

lawnjelly committed:
Add normalize_and_get_length() to Vector2 and Vector3

Length is calculated as part of the normalize routine. This is often desired, and also can be used to flag the condition where the Vector was too small to normalize (by returning zero length).

lawnjelly commented on core/math/vector2.h:
```diff
@@ -241,6 +242,17 @@ _FORCE_INLINE_ bool Vector2::operator!=(const Vector2 &p_vec2) const {
 
+real_t Vector2::normalize_and_get_length() {
+	real_t lengthsq = length_squared();
+	if (lengthsq >= (real_t)CMP_EPSILON) {
```
Ah yes, I'm expecting some discussion on the epsilon, and will need a bit of testing. I saw `CMP_EPSILON2` but will have to make sure it is ok for this use with 32 bit `real_t`.

AThousandShips replied:
I'm fully agnostic and the only difference with 64 bit would be the lower limit on magnitude

AThousandShips replied:
Though this could create an inconsistency with `is_normalized` as it uses `UNIT_EPSILON`

lawnjelly replied:
32 / 64 bit epsilon differences can be quite difficult.

Last time we discussed it (that I remember), reduz was in favour of keeping the 32 bit epsilons when compiling as 64 bit. This loses out on some potential accuracy in 64 bit, but if you use different epsilons you open up a massive surface of difficult to find bugs. Epsilon bugs are some of the _worst_ imo. :grin:

AThousandShips replied:
> Epsilon bugs are some of the worst imo.

Agreed

lawnjelly replied:
Ok after (I hope) some testing, I'm satisfied that `CMP_EPSILON2` should be okay at 32 bit, so I've pushed a version with this epsilon.

I tested in gdscript then in a c++ testbed. Of course there's always the chance I messed up, it's pretty easy to mess up, with math precision. :grin:

AThousandShips commented on doc/classes/Vector3.xml:
```diff
@@ -252,6 +252,13 @@
+			<return type="float" />
+			<description>
+				Scales the vector to unit length, and returns the original length.
+				If the original vector was too small to normalize, the length returned is [code]zero[/code].
```
There are also some issues with it because of access issues with GDScript, primitive types are non-mutable when accessed as properties, unless that has changed

lawnjelly replied:
If you can think of an easy way to return two values efficiently from bound functions, I'm happy to modify the PR.

If we don't want it to mutate the struct, we need to return the new Vector _and_ the length (or the new Vector _and_ success or failure).

Originally this was non-bound, but it is now also a potential candidate solution to the problem in #74852, and whichever solution we go with needs to be available from script.

Another non-mutating alternative would be a `normalized()` function that returns either the normalized vector or failure (maybe a Vector or NULL? can that work?).

lawnjelly replied:
Perhaps then for binding returning a variant that is either the normalized value or NULL is the way to go `try_normalized()` or something like that?

I can modify this particular PR to have this function unbound as before, but will be good to get some discussion going on how to solve the gdscript problem.

AThousandShips replied:
It could return an array or dictionary, somewhat cumbersome and inefficient but it gets it done

lawnjelly commented:
Closing in favour of #74912 , as it appears that modifying in place isn't good for bound functions.
